Added useIsKeyPressed (and global) hooks for key presses.

ezxyz opened the pull request:
The hooks will return a state (in boolean) that tracks whether the key is pressed. There is a global and non-global varient that requires a "ref".

multiSelectionMovable and singleSelectionMovable now uses the hooks.

Fixes #910

Commit:
Added useIsKeyPressed (and global) hooks for key presses.

The hooks will return a state (in boolean) that tracks whether the key is pressed. There is a global and non-global varient that requires a "ref".

multiSelectionMovable and singleSelectionMovable now uses the hooks.

Commit:
Update singleSelectionMovable.js

Commit:
Update singleSelectionMovable.js

using snapDisabled

Commit:
Create index.test.js

swissspidy reviewed (approved):
Overall LGTM

I'll look into the intermittent test failure(s)

dvoytenko commented on assets/src/edit-story/components/keyboard/index.js:
```diff
@@ -149,6 +149,31 @@ export function useGlobalKeyUpEffect(
+ * @param {Array|undefined} deps
+ * @return {boolean} Stateful boolean that tracks whether key is pressed.
+ */
+export function useGlobalIsKeyPressed(keyNameOrSpec, deps = undefined) {
```
You should just use your own `useIsKeyPressed()` here. You will just need to do `if (!globalRef.current) {...}` that other methods are doing. And I'll also request you put that `if` into a separate private method, e.g. `getGlobalRef()` or such.

ezxyz replied:
Done.

dvoytenko commented on assets/src/edit-story/components/keyboard/test/index.test.js:
```diff
@@ -0,0 +1,43 @@
+describe('keyboard/index.js', () => {
+  describe('useIsKeyPressed', () => {
+    it('should initialise and then register key up and down events', () => {
+      const { result } = renderHook(() => useIsKeyPressed(document, 'a'));
```
Let's create a fake element for this. Otherwise both: global and element-based could both easily work via `document` and we won't test the difference.

ezxyz replied:
Done.

Commit:
Modified tests and useGlobalIsKeyPressed to use useIsKeyPressed

Modified and added more tests to cover more edge cases, changed global key pressed to use useIsKeyPressed.

Added a helper function for globalref in keyboard.index.js.

miina commented on assets/src/edit-story/components/keyboard/index.js:
```diff
@@ -127,9 +147,7 @@ export function useGlobalKeyDownEffect(
-  if (!globalRef.current) {
-    globalRef.current = document;
-  }
+  const globalRef = getGlobalRef();
```
There is `const globalRef = createRef();` declared above already.

ezxyz replied:
Done.

miina commented on assets/src/edit-story/components/keyboard/index.js:
```diff
@@ -143,12 +161,20 @@ export function useGlobalKeyUpEffect(
-  if (!globalRef.current) {
-    globalRef.current = document;
-  }
+  const globalRef = getGlobalRef();
```
Ditto.

ezxyz replied:
Done.

miina commented on assets/src/edit-story/components/keyboard/index.js:
```diff
@@ -117,6 +124,19 @@ export function useKeyUpEffect(
   useKeyEffect(refOrNode, keyNameOrSpec, 'keyup', callback, deps);
```
^ Should the unused `useGlobalKeyUpEffect` be removed now? Doesn't look like it has any use anymore.

ezxyz replied:
Would it be strange to not have a useGlobalKeyUpEffect when we have a useKeyUpEffect and a useGlobalKeyDownEffect? I think it might not be consistent if someone were to use the hooks for something more custom than just listening to a keyPress.

However at the same time I'm not sure when it will get used, so I'm happy to remove this if you feel strongly. I think we should leave it in though.

dvoytenko replied:
I agree. Let's leave it for now.

miina replied:
I'd be in favor of removing unused code, especially since `useGlobalKeyUpEffect` is much less commonly used compared to `useGlobalKeyDownEffect`, and especially now that we have `useIsKeyPressed`, it could just be added back if it should become necessary.

However, looks like you're both in favor of keeping it though so marking as resolved, not feeling that strongly about it.

Commit:
removed duplicate globalRefs

dvoytenko commented on assets/src/edit-story/components/keyboard/index.js:
```diff
@@ -143,12 +160,20 @@ export function useGlobalKeyUpEffect(
+ * @return {boolean} Stateful boolean that tracks whether key is pressed.
+ */
+export function useGlobalIsKeyPressed(keyNameOrSpec, deps = undefined) {
+  const globalRef = getGlobalRef();
```
So which form did you decide to use? `setGlobalRef()` or `const globalRef = getGlobalRef()`? I only see `setGlobalRef()` declaration above.

ezxyz replied:
My bad, I accidently left this one in. Fixed.

Commit:
Update index.js

miina reviewed (approved):
LGTM!

swissspidy committed:
Merge branch 'master' into iskeypressedhook

swissspidy committed:
Merge branch 'master' into iskeypressedhook

swissspidy committed:
Fix lint errors
